src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -127,6 +128,216 @@ void CoreWorkerDirectTaskSubmitter::AddWorkerLeaseClient(
+  RAY_LOG(INFO) << "RETURN_WORKER " << addr.worker_id << " " << time_elapsed;
+
+  auto status =
+      lease_entry.lease_client->ReturnWorker(addr.port, addr.worker_id, was_error);
+  if (!status.ok()) {
+    RAY_LOG(ERROR) << "Error returning worker to raylet: " << status.ToString();
+  }
+  worker_to_lease_entry_.erase(addr);
+}
+
+void CoreWorkerDirectTaskSubmitter::StealWorkIfNeeded(
+    const rpc::WorkerAddress &thief_addr, bool was_error,
+    const SchedulingKey &scheduling_key,
+    const google::protobuf::RepeatedPtrField<rpc::ResourceMapEntry> &assigned_resources) {
+  if (!work_stealing_enabled_) {
+    RAY_LOG(DEBUG) << "Work stealing is not enabled, so we return the worker "
+                   << thief_addr.worker_id << " without stealing";
+    ReturnWorker(thief_addr, was_error, scheduling_key);
+    return;
+  }
+
+  RAY_LOG(DEBUG) << "Stealing work now! Thief is worker: " << thief_addr.worker_id;
+
+  auto &thief_entry = worker_to_lease_entry_[thief_addr];
+  RAY_CHECK(thief_entry.lease_client);
+  RAY_CHECK(thief_entry.tasks_in_flight == 0);
+  RAY_CHECK(thief_entry.stealable_tasks.size() == 0);
+
+  auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+
+  // Start looking for a suitable victim
+
+  if (scheduling_key_entry.active_workers.size() <= 1) {
+    RAY_LOG(DEBUG) << "Could not find a suitable victim for stealing! Returning worker "
+                   << thief_addr.worker_id;
+    ReturnWorker(thief_addr, was_error, scheduling_key);
+    return;
+  }
+  RAY_CHECK(scheduling_key_entry.active_workers.size() > 1);
+
+  absl::flat_hash_set<rpc::WorkerAddress>::iterator victim_it =
+      scheduling_key_entry.active_workers.begin();
+
+  for (absl::flat_hash_set<rpc::WorkerAddress>::iterator candidate_it =
+           scheduling_key_entry.active_workers.begin();
+       candidate_it != scheduling_key_entry.active_workers.end(); candidate_it++) {
+    rpc::WorkerAddress victim_addr = *victim_it;
+    rpc::WorkerAddress candidate_addr = *candidate_it;
+    RAY_CHECK(worker_to_lease_entry_.find(victim_addr) != worker_to_lease_entry_.end());
+    auto &victim_lease_entry = worker_to_lease_entry_[victim_addr];
+    auto &candidate_lease_entry = worker_to_lease_entry_[candidate_addr];
+
+    RAY_LOG(DEBUG) << "Current victim: " << victim_addr.worker_id << " with "
+                   << victim_lease_entry.tasks_in_flight << " tasks in flight";
+    RAY_LOG(DEBUG) << "Current candidate: " << candidate_addr.worker_id << " with "
+                   << candidate_lease_entry.tasks_in_flight << " tasks in flight";
+
+    if (victim_addr.worker_id == thief_addr.worker_id ||
+        ((candidate_lease_entry.stealable_tasks.size() >
+          victim_lease_entry.stealable_tasks.size()) &&
+         candidate_addr.worker_id != thief_addr.worker_id)) {
+      victim_it = candidate_it;
+    }
+  }
+  rpc::WorkerAddress victim_addr = *victim_it;
+
+  // Check that the victim is a suitable one
```

Comment:
I think that we could simplify these checks to just checking if the victim has more than one task in flight. The thief should have 0 so that will also make sure the worker doesn't steal from itself.